### ISSUE TITLE
HTML version?

### ISSUE THREAD

**Issue by AvyChanna** [vEnhance/napkin#116]
First of all, thank you for this amazing book(?). I wish I had known about napkin before my calculus classes.

I was wondering if there was a possibility for an HTML version for the same.
This would help in 2 ways -

- Permalink-ing to chapters/sections/theorems/... via links.
- Directly hosting the html copy on github.io for speedy reference.(like a gitbook/readthedocs)

This could've been as easy as `htlatex Napkin.tex` but I was not able to build it successfully.
The Travis CI uses some specific versions for everything (Ubuntu bionic, texlive ppa) and I was not able to build the same on my config.

(I expect everything but the diagrams to work well in HTML. Will try to reproduce the build environment on a Virtual Machine tomorrow, hope it succeeds.)

**Comment by vEnhance** [comment]
This is a nice idea. If you manage to get it to work, I would definitely be happy to take the pull request. I may also look into it myself, but it might not be any time soon.

**Comment by mikepierce** [comment]
I too am interested in this. I quite like the idea of moving mathematical writing online into webpages instead of in PDFs

**Comment by AvyChanna** [comment]
I was able to build PDF on Ubuntu 18.04.5. Found out, cleveref doesn't work well with htlatex.

```
) (/usr/share/texlive/texmf-dist/tex/latex/hyperref/htex4ht.def)
(/usr/share/texlive/texmf-dist/tex/latex/cleveref/cleveref.sty
! Undefined control sequence.
<argument> ...esult \relax \else \cref@addtoreset 
                                                  {parentequation}{\cref@res...
l.2954 }{}
          %  end of \@ifpackageloaded{amsmath}
? 
! Emergency stop.
<argument> ...esult \relax \else \cref@addtoreset 
                                                  {parentequation}{\cref@res...
l.2954 }{}
          %  end of \@ifpackageloaded{amsmath}
No pages of output.
Transcript written on Napkin.log.
This is pdfTeX, Version 3.14159265-2.6-1.40.20 (TeX Live 2019/Debian) (preloaded format=latex)
```

During testing, I removed `cleveref` and put a dummy command `\newcommand{\Cref}[1]{\$$1$}`. This resulted in a stack full error. Also, I found a tex.stackexchange answer to [cleveref + htlatex](https://tex.stackexchange.com/a/220345). This, plus using  `\AtBeginDocument{\usepackage{[nameinlink]cleveref}}` to load cleveref seems to be a step in the right direction. Still, the stack preblem persists.
Why? I don't know. Maybe infinite recursion? I'll have to read some more to get to find out. I'll try to increase stack size, maybe that would work.
```
! TeX capacity exceeded, sorry [input stack size=5000].
```

**Comment by AvyChanna** [comment]
Okay. I've identified the main issues. Currently, there are 2 problems with HTML generation.
- Diagrams result in errors. `tikz-cd` has its own `diagrams.4ht` file. `diagrams.sty` does not. These contain all the configurations for htlatex/make4ht.
- The version of LaTex is too old. Last update was a year ago. Much has changed in LaTeX core and TeX4ht since then.(Maybe it's time to update Travis CI build script).

Putting this issue on hold till the above 2 are resolved. Because my LaTex skills are not too good, I can only do the second one.

**Comment by vEnhance** [comment]
Sweet, glad you got it mostly working!

I'm hoping to eliminate `diagrams.sty` completely over time (but have been putting that off for nine months now), mentioned in #77. Once that is done, the integration here might become a little smoother.

I'm curious whether we think this can be made to work with the CI, to generate automatically?

**Comment by AvyChanna** [comment]
Possibly, but I'd advice against it.

Firstly, texlive version in current CI build is a whole major release behind(Texlive is "frozen" once a year).
You can get around this by building from source. You'll also need the most recent release of ghistscript to get that latest version of asympyote to run. GS had some internal fixes which change the way PS extensions are used(like transparency).

Secondly, build job is too lengthy for CI. Mine took forever. And [Travis CI is introducing a new price model](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing)

**Comment by i-ky** [comment]
Travis CI is still free for open source projects. They had to change their rules (blame cryptocurrency miners) and now you have to apply for your build time instead of getting unlimited access straight away. Quoting the link provided above:
> We will be offering an allotment of OSS minutes that will be reviewed and allocated on a case by case basis. Should you want to apply for these credits please open a request with Travis CI support stating that you’d like to be considered for the OSS allotment.

You will have to apply anyway since credits they give you for free trial will expire sooner or later.

**Comment by vEnhance** [comment]
> Possibly, but I'd advice against it.

Thanks for the suggestion. In that case, my thought is that, once #77 is resolved (which admittedly could be a while), then I can create a static version on my website, using the commands you suggest, that is periodically updated every so often. Won't be as fresh as the main PDF but I haven't had time for the last few years to do more than trivial edits anyways.

**Comment by AvyChanna** [comment]
Sorry. I was(and will be) busy with academics and had to put this aside for the moment.

There are some corrections required for HTML output. Current maths output is bad and needs to be replaced with MathML/SVG/Katex. And, I'll have to repair/replace fixed units of length. PDF is expected to look the same on all devices, but HTML is more "fluid" in nature.

I won't be able to work on it until April. If someone else is interested, they can work on it in the meantime.

**Comment by maparent** [comment]
Did a bit of work on this in https://github.com/maparent/napkin/tree/htlatex . I can now apply `htlatex Napkin.tex mathml,svg` to most of Napkin except three chapters (quantum/circuits, quantum/shor and measure/swapsum). Images rendering is also still problematic, but it's a step.

**Comment by maparent** [comment]
More info: I have removed the `settings.outformat` on asymptote on my branch, so it is easier to create the svg version of the asymptote figures. However, this breaks the mklatex process, so it is (obviously) provisional. The procedure right now is to run mklatex once to generate the `asy/*.asy` files, then use `asy` directly to generate all .svg and .pdf from those (doing it with a shell loop for now), then run latexmk to completion, and finally htlatex (and tex4ebook) can complete. I have disastrous results with dvisvgm on OSX, so most tikz images  still do not render well.

**Comment by maparent** [comment]
thought of externalizing the Tikz images, but ran into [this issue](https://tex.stackexchange.com/questions/171931/are-the-tikz-libraries-cd-and-external-incompatible-with-one-another). The link gives a lot of workarounds, none exactly lightweight. Curious if this is worth pursuing.
